proxy/src/drain.rs

```diff
@@ -0,0 +1,140 @@
+use std::mem;
+
+use futures::{Async, Future, Poll, Stream};
+use futures::future::Shared;
+use futures::sync::{mpsc, oneshot};
+
+/// Creates a drain channel.
+///
+/// The `Signal` is used to start a drain, and the `Watch` will be notified
+/// when a drain is signaled.
+pub fn channel() -> (Signal, Watch) {
+    let (tx, rx) = oneshot::channel();
+    let (drained_tx, drained_rx) = mpsc::channel(0);
+    (
+        Signal {
+            drained_rx,
+            tx,
+        },
+        Watch {
+            drained_tx,
+            rx: rx.shared(),
+        },
+    )
+}
+
+/// Send a drain command to all watchers.
+///
+/// When a drain is started, this returns a `Drained` future which resolves
+/// when all `Watch`ers have been dropped.
+#[derive(Debug)]
+pub struct Signal {
+    drained_rx: mpsc::Receiver<Never>,
+    tx: oneshot::Sender<()>,
+}
+
+/// Watch for a drain command.
+///
+/// This wraps another future and callback to be called when drain is triggered.
+#[derive(Clone, Debug)]
+pub struct Watch {
+    drained_tx: mpsc::Sender<Never>,
+    rx: Shared<oneshot::Receiver<()>>,
+}
+
+/// The wrapped watching `Future`.
+#[derive(Debug)]
+pub struct Watching<A, F> {
+    future: A,
+    state: State<F>,
+    watch: Watch,
+}
+
+#[derive(Debug)]
+enum State<F> {
+    Watch(F),
+    Draining,
+}
+
+#[derive(Debug)]
+enum Never {}
+
+/// A future that resolves when all `Watch`ers have been dropped (drained).
+pub struct Drained {
+    drained_rx: mpsc::Receiver<Never>,
+}
+
+// ===== impl Signal =====
+
+impl Signal {
+    pub fn drain(self) -> Drained {
+        let _ = self.tx.send(());
+        Drained {
+            drained_rx: self.drained_rx,
+        }
+    }
+}
+
+// ===== impl Watch =====
+
+impl Watch {
+    pub fn watch<A, F>(self, future: A, on_drain: F) -> Watching<A, F>
```

Comment:
perhaps add a brief docstring here to the effect of _Wraps `future` so that `on_drain` is called with it whena drain is signaled._